daemon/settings.c

```diff
@@ -962,8 +1004,21 @@ appsettings_set_permissions(appsettings_t *self, const stringset_t *permissions)
             appsettings_notify_change(self);
         }
 
-        if( reset_allow )
-            appsettings_set_allowed(self, APP_ALLOWED_UNSET);
+        if( reset_allow ) {
+            /* If launching is allowlisted, instead of reseting permissions
+             * we should only extend with new permissions to avoid allowing
+             * previously denied permissions.
```

Comment:
Kind of describes how I felt about this while implementing. This idea of keeping the previous value of autogrant in settings data sounds like a good idea. I'll explore that.

Reply:
Updated but at the same time dropped this whole extending thing out of the changes. I'm not yet sure how I should handle that but I do like the idea that if APP_GRANT_LAUNCH means permissions is allowed by default, then that extends also to new permissions. `appsettings_set_permissions()` is probably the wrong place, but also `appsettings_set_granted()` doesn't work as it changes `self->ast_permissions` before autogrant value is read. Perhaps moving it after autogrant setup in `appsettings_decode()` would allow it to work. I'll look at that tomorrow, now it's too late to improve further.

Reply:
Fixed that part. Now it should be a lot simpler and leaner implementation that doesn't have weird double states so it's easier to reason about and test.